src/io/filesystem/illegal_filename_characters.h

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (C) 2021 by the Widelands Development Team
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+
+#ifndef WL_IO_FILESYSTEM_ILLEGAL_FILENAME_CHARACTERS_H
+#define WL_IO_FILESYSTEM_ILLEGAL_FILENAME_CHARACTERS_H
+
+namespace {
+// Characters that are allowed in filenames, but not at the beginning
+static const std::vector<std::string> illegal_filename_starting_characters{
```

Comment:
Header files shouldn't contain anonymous namespaces or definitions of non-trivial types. Remove the namespace, declare the vector as 
`extern const std::vector<std::string> illegal_filename_starting_characters;`,
and move its definition (`std::vector<std::string> illegal_filename_starting_characters{...};`) to a .cc file. Same for other static variables in headers.

Reply:
I tried to do this in `filesystem.h` (leaving the definition in `filesystem.cc`, but it failed in final linking. I've just tried again with adding an `illegal_filename_characters.cc`, and it fails just the same:

```
/usr/bin/ld: /tmp/ccTz6BSq.ltrans7.ltrans.o: in function `FileSystem::is_legal_filename(std::__cxx11::basic_string<char,
std::char_traits<char>, std::allocator<char> > const&)':
<artificial>:(.text+0xa10e): undefined reference to `illegal_filename_starting_characters[abi:cxx11]'
/usr/bin/ld: <artificial>:(.text+0xa153): undefined reference to `illegal_filename_characters[abi:cxx11]'
/usr/bin/ld: /tmp/ccTz6BSq.ltrans18.ltrans.o: in function `FileSystemHelper::illegal_filename_tooltip[abi:cxx11]()':
<artificial>:(.text+0x2be44): undefined reference to `illegal_filename_starting_characters[abi:cxx11]'
/usr/bin/ld: <artificial>:(.text+0x2c20b): undefined reference to `illegal_filename_characters[abi:cxx11]'
```

Reply:
Maybe this is related to why `find_unused_includes.py` thinks the .cc-s don't need `illegal_filename_characters.h`?

Reply:
[patch.txt](https://github.com/widelands/widelands/files/7739626/patch.txt)
this should fix both the linker errors and the include script

Reply:
Indeed... But why is it different when the definitions are in `...tooltips.cc` than when in `filesystem.cc`? Is it something like when both the decl. and def. get linked into `libio_filesystem.a`, the extern doesn't apply anymore?

But this way `filesystem.cc` still depends on `...tooltips.cc`, and even does so silently.

Reply:
hmmm, this seems to be the answer: https://stackoverflow.com/questions/875655/linking-extern-variables-in-c

So it can be left alone in filesystem.cc, there's no need for illegal_chars.h, and it should be declared as extern in tooltip.cc...

Trying it now...

Reply:
OK, turns out I can move all of it to the new lib. I would have sworn that I had checked that `is_legal_filename()` or the character lists were used by the filesystem code...